File: app/src/main/java/tag_them/metoothanks/CanvasView.java
package tag_them.metoothanks;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.os.SystemClock;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.util.AttributeSet;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;

import tag_them.metoothanks.items.Image;
import tag_them.metoothanks.items.Item;
import tag_them.metoothanks.items.Text;

import java.util.ArrayList;

import static tag_them.metoothanks.Utils.reverse;
import static tag_them.metoothanks.Utils.swapWithNextItem;
import static tag_them.metoothanks.Utils.swapWithPrevItem;


public class CanvasView extends View {
	private ArrayList<Item> items = new ArrayList<>();
	private Item selectedItem;
	
	private Toolbar itemToolbar;
	
	// starting point of a drag gesture
	private int gripX, gripY,
		gripRight, gripBottom;
	private float heightWidthratio;
	
	// time of last touch
	private long prevTouchTime = SystemClock.elapsedRealtime();
	
	private static final int ACTION_MOVE = 1,
		ACTION_RESIZE = 2;
	private int lastAction, grippedEdge;
	
	// if a touch is registered less than 101 milliseconds
	// since the last touch, it will be treated as a drag/slide
	private final int MAX_ELAPSED_TIME_TO_DRAG = 100;
	
	public CanvasView(Context context) {
		super(context);
		setDrawingCacheEnabled(true);
	}
	
	public CanvasView(Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
		setDrawingCacheEnabled(true);
	}
	
	public void addImage(Bitmap bitmap) {
		// resizing the image in case it's bigger than the view/canvas
		float ratio;
		float viewW = getMeasuredWidth(),
			viewH = getMeasuredHeight();
		Bitmap finalBitmap = bitmap;
		if (finalBitmap.getWidth() > viewW) {
			ratio = viewW / finalBitmap.getWidth();
			finalBitmap = Bitmap.createScaledBitmap(bitmap, getMeasuredWidth(), (int) (finalBitmap.getHeight() * ratio), false);
		}
		if (finalBitmap.getHeight() > viewH) {
			ratio = viewH / finalBitmap.getHeight();
			finalBitmap = Bitmap.createScaledBitmap(bitmap, (int) (finalBitmap.getWidth() * ratio), getMeasuredHeight(), false);
		}
		
		// registering the image
		Image newImage = new Image(finalBitmap, this);
		items.add(newImage);
		selectItem(newImage);
		invalidate();
	}
	
	public void addText(String text) {
		Text newText = new Text(text, getMeasuredWidth(), this);
		items.add(newText);
		selectItem(newText);
	}
	
	public void removeSelectedItem() {
		items.remove(selectedItem);
		selectNone();
		invalidate();
	}
	
	@Override
	protected void onDraw(Canvas canvas) {
		for (Item i : items)
			i.draw(canvas);
	}
	
	@Override
	public boolean onTouchEvent(MotionEvent event) {
		int x = (int) event.getX(),
			y = (int) event.getY();
		
		// looking for a picture that got touched;
		// since the first pictures in the array get drawn first,
		// we iterate over the array from its end in order to get the
		// furthest picture (picture in the front)
		for (Item i : reverse(items)) {
			if (SystemClock.elapsedRealtime() - prevTouchTime < MAX_ELAPSED_TIME_TO_DRAG
				&& selectedItem != null) {
				switch (lastAction) {
					case ACTION_MOVE:
						selectedItem.move(x - gripX, y - gripY);
						break;
					
					case ACTION_RESIZE:
						switch (grippedEdge) {
							case Item.LEFT_EDGE:
								selectedItem.setLeft(x - gripX);
								break;
							case Item.TOP_EDGE:
								selectedItem.setTop(y - gripY);
								break;
							case Item.RIGHT_EDGE:
								selectedItem.setRight(x - gripRight);
								break;
							case Item.BOTTOM_EDGE:
								selectedItem.setBottom(y - gripBottom);
								break;
							case Item.TOP_LEFT_CORNER:
								selectedItem.setLeft(x - gripX);
								selectedItem.setTop(selectedItem.getBottom() - (int) (selectedItem.getWidth() * heightWidthratio));
								break;
							case Item.TOP_RIGHT_CORNER:
								selectedItem.setRight(x - gripRight);
								selectedItem.setTop(selectedItem.getBottom() - (int) (selectedItem.getWidth() * heightWidthratio));
								break;
							case Item.BOTTOM_RIGHT_CORNER:
								selectedItem.setRight(x - gripRight);
								selectedItem.setBottom(selectedItem.getTop() + (int) (selectedItem.getWidth() * heightWidthratio));
								break;
							case Item.BOTTOM_LEFT_CORNER:
								selectedItem.setLeft(x - gripX);
								selectedItem.setBottom(selectedItem.getTop() + (int) (selectedItem.getWidth() * heightWidthratio));
								break;
						}
						break;
				}
			} else {
				heightWidthratio = (float) i.getHeight() / (float) i.getWidth();
				gripX = x - i.getLeft();
				gripY = y - i.getTop();
				gripRight = x - i.getRight();
				gripBottom = y - i.getBottom();
				
				grippedEdge = i.getTouchedEdge(x, y);
				if (grippedEdge != Item.EDGE_NONE) {
					selectItem(i);
					lastAction = ACTION_RESIZE;
				} else {
					if (i.isTouched(x, y)) {
						selectItem(i);
						lastAction = ACTION_MOVE;
					} else {
						selectNone();
					}
				}
			}
			
			prevTouchTime = SystemClock.elapsedRealtime();
		}
		
		invalidate();
		
		return true;
	}
	
	public void prepareToExport(Canvas canvas) {
		selectNone();
		canvas.drawColor(Color.WHITE);
		draw(canvas);
	}
	
	private void selectNone() {
		itemToolbar.setVisibility(Toolbar.INVISIBLE);
		
		if (selectedItem != null)
			selectedItem.deselect();
		selectedItem = null;
	}
	
	private void selectItem(Item item) {
//		// swapping with the last image in the array
//		// so the selected image will be drawn at the front
//		Collections.swap(images, images.indexOf(image), images.size() - 1);

//		 add a delete option to the menu bar
		
		itemToolbar.getMenu().clear();
		itemToolbar.inflateMenu(R.menu.item_menu);
		itemToolbar.inflateMenu(item.getMenuID());
		itemToolbar.setOnMenuItemClickListener(getOnMenuItemClickListener(item.getClickListener()));
		itemToolbar.setVisibility(Toolbar.VISIBLE);
		
		if (selectedItem != null)
			selectedItem.deselect();
		selectedItem = item;
		selectedItem.select();
	}
	
	public void setItemToolbar(Toolbar itemToolbar) {
		this.itemToolbar = itemToolbar;
	}
	
	public Toolbar.OnMenuItemClickListener getOnMenuItemClickListener(final ClickListener additionalListener) {
		return item -> {
			int id = item.getItemId();
			switch (id) {
				case R.id.action_item_delete:
					removeSelectedItem();
					invalidate();
					return true;
					
				case R.id.action_item_move_forward:
					swapWithNextItem(selectedItem, items);
					invalidate();
					return true;
					
				case R.id.action_item_move_backward:
					swapWithPrevItem(selectedItem, items);
					invalidate();
					return true;
					
				default:
					additionalListener.handleClick(item);
			}
			return false;
		};
	}
	
	public interface ClickListener {
		boolean handleClick(MenuItem item);
	}
}

File: app/src/main/java/org/itiswednesday/metoothanks/CanvasView.kt
package org.itiswednesday.metoothanks

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.os.SystemClock
import android.support.v7.widget.Toolbar
import android.util.AttributeSet
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import org.itiswednesday.metoothanks.activities.Edit
import org.itiswednesday.metoothanks.items.Image
import org.itiswednesday.metoothanks.items.Item
import org.itiswednesday.metoothanks.items.Text
import java.io.Serializable
import android.graphics.DashPathEffect



const val CORNER_RADIUS = 25f
const val EDGE_WIDTH = 10f

class CanvasView : View, Serializable {
    private var selectedItem: Item? = null
    private val items = ArrayList<Item>()

    constructor(context: Context?) : super(context)

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)

    private val framePaint = Paint().apply {
        style = Paint.Style.STROKE
        strokeWidth = EDGE_WIDTH
        color = fetchColor(context, R.attr.colorPrimary)
        pathEffect = DashPathEffect(floatArrayOf(50f, 50f), 0f)
    }

    override fun onDraw(canvas: Canvas) {
        items.forEach {
            it.draw(canvas)
        }

        if (selectedItem != null)
            canvas.drawRoundRect(selectedItem!!.bounds, CORNER_RADIUS, CORNER_RADIUS, framePaint)
    }


    fun addImage(bitmap: Bitmap, center: Boolean = false) =
    // fitBitmap – resizing the image in case it's bigger than the screen
            Image(fitBitmap(bitmap, width, height), hostView = this).apply {
                if (center)
                    move(this@CanvasView.width / 2 - this.width / 2,
                         this@CanvasView.height / 2 - this.height / 2)
            }.addToItems()

    fun addText(text: String) = Text(text, width, hostView = this).addToItems()


    // if a touch is registered less than 101 milliseconds
    // since the last touch, it will be treated as a drag/slide
    private val MAX_ELAPSED_TIME_TO_DRAG = 100

    // last time the canvas was touched with one pointer
    private var prevMoveTouchTime = SystemClock.elapsedRealtime()

    // last time the canvas was touched with two pointers
    private var prevResizeTouchTime = SystemClock.elapsedRealtime()

    // the distance between the X coordinate of the pointer and the item's left edge (used for dragging)
    private var gripDistanceFromLeftEdge = 0

    // the distance between the Y coordinate of the pointer and the item's top edge (used for dragging)
    private var gripDistanceFromTopEdge = 0

    // pointers' coordinates at the beginning of a resize
    private var pointersGrip = PointPair({ Point() })

    // bounds value at the beginning of the resize
    private var originalBounds = RectF()

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent?): Boolean {
        if (event == null) return false

        val pointerCount = event.pointerCount

        when (pointerCount) {
            1 -> {
                val x = event.x.toInt()
                val y = event.y.toInt()

                // looking for a picture that got touched;
                // since the first pictures in the array get drawn first,
                // we iterate over the array from its end in order to get the
                // furthest picture (picture in the front)
                for (item in items.reversed()) {
                    if (SystemClock.elapsedRealtime() -
                            prevMoveTouchTime < MAX_ELAPSED_TIME_TO_DRAG && selectedItem != null)
                        selectedItem?.move(
                                x - gripDistanceFromLeftEdge, y - gripDistanceFromTopEdge)
                    else {
                        selectNone()
                        if (item.touched(x, y)) {
                            gripDistanceFromLeftEdge = x - item.left
                            gripDistanceFromTopEdge = y - item.top

                            item.select()
                        }
                    }

                    prevMoveTouchTime = SystemClock.elapsedRealtime()
                }
            }
            2 -> {
                if (SystemClock.elapsedRealtime() -
                        prevResizeTouchTime < MAX_ELAPSED_TIME_TO_DRAG) {
                    val pointers = PointPair(
                            init = { Point(event.getX(it).toInt(), event.getY(it).toInt()) })

                    selectedItem?.resize(pointers, pointersGrip, originalBounds)
                } else {
                    pointersGrip = PointPair {
                        Point(event.getX(it).toInt(), event.getY(it).toInt())
                    }

                    selectedItem?.let { originalBounds = it.bounds }
                }

                prevResizeTouchTime = SystemClock.elapsedRealtime()
            }
        }

        invalidate()

        return true
    }

    private fun Item.select() {
        selectedItem = this

        val clickListener = { item: MenuItem ->
            when (item.itemId) {
                R.id.action_item_delete        -> selectedItem!!.removeFromItems()
                R.id.action_item_move_forward  -> items.swapWithNextItem(selectedItem!!)
                R.id.action_item_move_backward -> items.swapWithPrevItem(selectedItem!!)
                else                           -> handleMenuItemClick(item)
            }

            true
        }

        with(itemToolbar) {
            menu.clear()
            inflateMenu(R.menu.item_menu)
            inflateMenu(itemMenuID)
            setOnMenuItemClickListener(clickListener)
            visibility = Toolbar.VISIBLE
        }

        invalidate()
    }


    fun selectNone() {
        selectedItem = null
        itemToolbar.menu.clear()
    }

    private fun Item.addToItems() {
        items.add(this)
        this.select()
        invalidate()
    }

    private fun Item.removeFromItems() {
        items.remove(this)
        selectNone()
        invalidate()
    }

    private val itemToolbar
        get() = (context as Edit).findViewById<Toolbar>(R.id.item_toolbar)
}